commit 0b053e78b5990cd01e7169ee5bd2bb8e4045deea
Author: David Rowley <drowley@postgresql.org>
Date:   Thu Oct 5 20:30:47 2023 +1300

    Fix memory leak in Memoize code
    
    Ensure we switch to the per-tuple memory context to prevent any memory
    leaks of detoasted Datums in MemoizeHash_hash() and MemoizeHash_equal().
    
    Reported-by: Orlov Aleksej
    Author: Orlov Aleksej, David Rowley
    Discussion: https://postgr.es/m/83281eed63c74e4f940317186372abfd%40cft.ru
    Backpatch-through: 14, where Memoize was added

diff --git a/src/backend/executor/nodeMemoize.c b/src/backend/executor/nodeMemoize.c
index 94bf4792873..1085b3c79dd 100644
--- a/src/backend/executor/nodeMemoize.c
+++ b/src/backend/executor/nodeMemoize.c
@@ -156,14 +156,18 @@ static bool MemoizeHash_equal(struct memoize_hash *tb,
  */
 static uint32
 MemoizeHash_hash(struct memoize_hash *tb, const MemoizeKey *key)
 {
 	MemoizeState *mstate = (MemoizeState *) tb->private_data;
+	ExprContext *econtext = mstate->ss.ps.ps_ExprContext;
+	MemoryContext oldcontext;
 	TupleTableSlot *pslot = mstate->probeslot;
 	uint32		hashkey = 0;
 	int			numkeys = mstate->nkeys;
 
+	oldcontext = MemoryContextSwitchTo(econtext->ecxt_per_tuple_memory);
+
 	if (mstate->binary_mode)
 	{
 		for (int i = 0; i < numkeys; i++)
 		{
 			/* combine successive hashkeys by rotating */
@@ -201,10 +205,12 @@ MemoizeHash_hash(struct memoize_hash *tb, const MemoizeKey *key)
 				hashkey ^= hkey;
 			}
 		}
 	}
 
+	ResetExprContext(econtext);
+	MemoryContextSwitchTo(oldcontext);
 	return murmurhash32(hashkey);
 }
 
 /*
  * MemoizeHash_equal
@@ -224,33 +230,46 @@ MemoizeHash_equal(struct memoize_hash *tb, const MemoizeKey *key1,
 	/* probeslot should have already been prepared by prepare_probe_slot() */
 	ExecStoreMinimalTuple(key1->params, tslot, false);
 
 	if (mstate->binary_mode)
 	{
+		MemoryContext oldcontext;
 		int			numkeys = mstate->nkeys;
+		bool		match = true;
+
+		oldcontext = MemoryContextSwitchTo(econtext->ecxt_per_tuple_memory);
 
 		slot_getallattrs(tslot);
 		slot_getallattrs(pslot);
 
 		for (int i = 0; i < numkeys; i++)
 		{
 			FormData_pg_attribute *attr;
 
 			if (tslot->tts_isnull[i] != pslot->tts_isnull[i])
-				return false;
+			{
+				match = false;
+				break;
+			}
 
 			/* both NULL? they're equal */
 			if (tslot->tts_isnull[i])
 				continue;
 
 			/* perform binary comparison on the two datums */
 			attr = &tslot->tts_tupleDescriptor->attrs[i];
 			if (!datum_image_eq(tslot->tts_values[i], pslot->tts_values[i],
 								attr->attbyval, attr->attlen))
-				return false;
+			{
+				match = false;
+				break;
+			}
 		}
-		return true;
+
+		ResetExprContext(econtext);
+		MemoryContextSwitchTo(oldcontext);
+		return match;
 	}
 	else
 	{
 		econtext->ecxt_innertuple = tslot;
 		econtext->ecxt_outertuple = pslot;
